### PR TITLE
add docker compose

### PR DESCRIPTION
**Pull request by siosw**
# Goals of PR

adds a docker compose script that brings up all components of a local devnet

## How to run
copy `config/deployments/local_docker` into `workspace`
`docker compose -f docker/docker-compose-test.yml build`
`docker compose -f docker/docker-compose-test.yml up`

## Validator error
this is fixed by #223 
There still is some issue with running the validator
This error can be reproduced by setting the validator flags in the `start_sidecar` script
```
docker-sidecar-1  | ERROR[10-30|16:40:28.509] "func1:80 | Failed triggering callback: received two finalized headers for block_num=0 with hashes: 0x6ec179995fc1d338c90173c9f70a29ebadf30a746dd9d2dbb178b5492103b959 and 0x0000000000000000000000000000000000000000000000000000000000000000
```

## TODO
- [x] fix configure rollup script in dockerized L1
- [ ] `start_l1` should still work as a standalone script
- [x] setup intra container networking
- [ ] remove hardcoded timeouts in `start_*` scripts

### CODE REVIEW

**Review by macwis — APPROVED** [review]
works, IMHO we should merge this ASAP

File: sbin/deploy_l1_contracts.sh [hunk 1]
```diff
@@ -51,14 +51,18 @@ relpath () {
...
-	read -r -p "Overwrite $1 with a new file? [y/N] " response
-	if [[ "$response" =~ ^([yY][eE][sS]|[yY])$ ]]; then
-	    rm $1
-	else
-	    exit
-	fi
+  if [ -z "$PS1" ]; then
+    echo "no interactive shell, ignoring guard..."
+    return 0
+  fi
```
Comment (ujvl) [inline sbin/deploy_l1_contracts.sh:57]: imo we can get rid of this and just do `yes | sbin/deploy_l1_contracts.sh`

File: sbin/start_l1.sh [hunk 2]
```diff
@@ -83,9 +55,16 @@ else
     exit 1
 fi
 
+# Optionally deploy the contracts
+if [ "$L1_DEPLOY" = "true" ]; then
```
Comment (ujvl) [inline sbin/start_l1.sh:59]: any new parameters must be documented in `config/example`

File: sbin/start_sidecar.sh [hunk 3]
```diff
@@ -16,19 +9,32 @@ fi
...
+    --rollup.l1.rollup-genesis-block $GENESIS_L1_BLOCK_NUM \
+    --rollup.l2.endpoint $L2_ENDPOINT \
+    --rollup.l2.chainid $NETWORK_ID \
+    --rollup.sequencer.addr $SEQUENCER_ADDR \
+    --sequencer.secret-key $SEQUENCER_SECRET_KEY \
+    --validator.secret-key $VALIDATOR_SECRET_KEY \
+"
+
+
+sleep 15
```
Comment (ujvl) [inline sbin/start_sidecar.sh:35]: this doesn't make sense to do here; the services should run as standalone components and not assume how long other services take to boot

**Review by ujvl — COMMENTED** [review]
post-merge comments to be addressed

File: contracts/scripts/config/create_config.ts [hunk 4]
```diff
@@ -51,7 +51,9 @@ export async function generateConfigFile(
   baseConfig.genesis.l1.number = l1Number;
   baseConfig.genesis.l2.hash = l2Hash;
   const genesis = JSON.parse(fs.readFileSync(genesisPath, "utf-8"));
-  baseConfig.genesis.l2_time = genesis.timestamp;
+  // TODO: why is the timestamp not being generated ed?
```
Comment (ujvl) [inline contracts/scripts/config/create_config.ts:54]: what does this todo mean?

File: sbin/create_rollup_config.sh
Comment (ujvl) [inline sbin/create_rollup_config.sh:1]: extraneous file..

File: sbin/start_l1.sh [hunk 5]
```diff
@@ -83,9 +55,16 @@ else
...
+    echo "deploying contracts..."
+    #npx hardhat deploy --network localhost
```
Comment (ujvl) [inline sbin/start_l1.sh:61]: avoid merging in commented lines like this

File: config/local_docker/.env
Comment (ujvl) [inline config/local_docker/.env:1]: what is this file for? is it overriding other dotenvs?

File: sbin/start_l1.sh [hunk 6]
```diff
@@ -83,9 +55,16 @@ else
...
+    bash $SBIN/deploy_l1_contracts.sh
+fi
+
 # Follow output
 if [ "$L1_STACK" = "geth" ]; then
-    docker logs l1_geth --follow
+    tail -f $PID
 elif [ "$L1_STACK" = "hardhat" ]; then
```
Comment (ujvl) [inline sbin/start_l1.sh:68]: you can merge these statements if they do the same thing

File: sbin/start_l1.sh [hunk 7]
```diff
@@ -83,9 +55,16 @@ else
     exit 1
 fi
 
+# Optionally deploy the contracts
+if [ "$L1_DEPLOY" = "true" ]; then
+    echo "deploying contracts..."
+    #npx hardhat deploy --network localhost
+    bash $SBIN/deploy_l1_contracts.sh
```
Comment (ujvl) [inline sbin/start_l1.sh:62]: where is $SBIN set? it looks like you removed it at the top?

File: sbin/start_l1.sh [hunk 8]
```diff
@@ -83,9 +55,16 @@ else
...
+fi
+
 # Follow output
 if [ "$L1_STACK" = "geth" ]; then
-    docker logs l1_geth --follow
+    tail -f $PID
```
Comment (ujvl) [inline sbin/start_l1.sh:67]: should this be `$L1_PID`?

File: sbin/start_l1.sh [hunk 9]
```diff
@@ -22,54 +15,33 @@ L1_HOST=`echo $L1_ENDPOINT | awk -F':' '{print substr($2, 3)}'`
...
-    for pid in "${PIDS[@]}"; do
-        echo "Killing $pid"
-        kill $pid
-    done
-}
-
-function ctrl_c() {
-    cleanup
-}
-##########################
```
Comment (ujvl) [inline sbin/start_l1.sh:42]: why remove this?

File: sbin/start_sp_geth.sh [hunk 10]
```diff
@@ -22,49 +22,43 @@ while getopts "$optspec" optchar; do
...
 echo "Using bin: $SP_GETH"
 
-# Check that the dotenv exists.
-ENV=".sp_geth.env"
-if ! test -f $ENV; then
-    echo "Expected dotenv at $ENV (does not exist)."
-    exit
-fi
-echo "Using dotenv: $ENV"
-. $ENV
```
Comment (ujvl) [inline sbin/start_sp_geth.sh:39]: why is this removed?

File: sbin/start_l1.sh [hunk 11]
```diff
@@ -1,11 +1,4 @@
-#!/bin/bash
-if [ -z $CONTRACTS_DIR ]; then
-    SBIN=`dirname $0`
-    SBIN="`cd "$SBIN"; pwd`"
-    . $SBIN/configure.sh
-fi
-echo "Using $CONTRACTS_DIR as deployment source"
```
Comment (ujvl) [inline sbin/start_l1.sh:7]: why is this removed?
